### PR TITLE
Project.load_data Handle multiplex metadata logic

### PR DESCRIPTION
**Pull request by avrohomgottlieb**
## Issue Number

https://github.com/AlexsLemonade/scpca-portal/issues/663

## Purpose/Implementation Notes

Following up from https://github.com/AlexsLemonade/scpca-portal/pull/652, (which dealt with single-cell and spatial metadata), this PR addresses the handling of multiplexed metadata, specifically:
- updating its sample and library metadata loading inside of `Project::load_samples_metadata` and `Project::load_libraries_metadata` respectively
- integrating `Project::combine_multiplexed_metadata` into the generic `Project::combine_metadata` method

## Types of changes

What types of changes does your code introduce?

- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Screenshots

N/A

### CODE REVIEW

**Comment by avrohomgottlieb** [comment]
Bug alert: in `get_library_metadata_keys`, modality/modalities kwarg don't agree between usage and implementation

**Comment by avrohomgottlieb** [comment]
Still to do:
- Adding of docstrings to newly added methods

Thinking about doing:
- More aggressive refactor of `Project::combine_metadata` to move out the key generation at the beginning of the method to its own method. Further down the line it would also be good to look at the `Project::get_sample_metadata_keys` and `Project::get_library_metadata_keys` methods to see if they can be rewritten in a way that's more declarative, and to update the usage of the modality/modalities tuple.
- Tucking of `Project::load_sample_metadata` into  `Project::load_library_metadata` in order to manage a single state of the `samples_metadata`->`updated_samples_metadata` dict.

**Comment by avrohomgottlieb** [comment]
> Bug alert: in `get_library_metadata_keys`, modality/modalities kwarg don't agree between usage and implementation. This is connected to the adding of `demux_cell_count_estimate` and `multiplexed_with` keys in `get_sample_metadata_keys` inside the general `excluded_keys` dict, as opposed to within the multiplexed block. When put within the multiplexed block, errors arise in test suite.

**Comment by avrohomgottlieb** [comment]
This PR also contributes to https://github.com/AlexsLemonade/scpca-portal/issues/638 by removing the mutations of the four multiplexed library metadata fields from `Project::get_samples`, and by moving them to `Project::load_libraries_metadata`.

File: api/scpca_portal/models/project.py [hunk 1]
```diff
@@ -745,6 +463,166 @@ def get_additional_terms(self):
...
+                    multiplexed_json = json.load(multiplexed_json_file)
+
+                multiplexed_json["scpca_library_id"] = multiplexed_json.pop("library_id")
+                multiplexed_json["scpca_sample_id"] = multiplexed_json.pop("sample_id")
+
+                multiplexed_libraries_metadata.append(multiplexed_json)
+
+        return multiplexed_libraries_metadata
+
+    def get_multiplexed_remaining_fields(self):
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:493]: Any reason not to call these aggregate fields?

File: api/scpca_portal/models/project.py [hunk 2]
```diff
@@ -745,6 +463,166 @@ def get_additional_terms(self):
...
+    ):
+        """
+        Returns a list of combined metadata of all samples that
+        the scpca_sample_id sample was multiplexed with.
+        The returned list is then immediately written to the open csv file in the caller method.
+        """
+        multiplexed_with_combined_metadata = []
+        multiplexed_sample_ids = sorted(multiplexed_with_mapping[scpca_sample_id])
+
+        multiplexed_with_samples_metadata_filtered_keys = {
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:590]: This name is pretty hard to parse without looking at how it is used. I think an inline comment describing the variable's value would be prudent here.

File: api/scpca_portal/models/project.py [hunk 3]
```diff
@@ -750,12 +752,31 @@ def get_metadata_field_names(self, columns, modality):
...
+        Returns a dictionary which maps sample ids to a set of associated library ids.
+        """
+        # sample_libraries_id_mapping = {sample_id: set() for sample_id in sample_ids}
+        sample_libraries_id_mapping = {}
+
         for library_metadata in libraries_metadata:
-            sample_libraries_id_mapping[library_metadata["scpca_sample_id"]].add(
-                library_metadata["scpca_library_id"]
-            )
+            isMultiplexed = "demux_samples" in library_metadata
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:763]: `is_multiplexed`

File: api/scpca_portal/models/project.py [hunk 4]
```diff
@@ -750,12 +752,31 @@ def get_metadata_field_names(self, columns, modality):
...
+
+            if not isMultiplexed:
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:765]: can we reverse the order of these cases?

File: api/scpca_portal/models/project.py [hunk 5]
```diff
@@ -1213,25 +1213,22 @@ def combine_metadata(self, updated_samples_metadata, libraries_metadata, sample_
                 modality=modality,
             )
 
-            sample_libraries_mapping = self.get_sample_libraries_mapping(
-                libraries_metadata[modality]
+            pre_filtered_samples_metadata = (
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:1216]: Would you be alright with `unfiltered_samples_metadata` or `input_samples_metadata`

**Review by davidsmejia — COMMENTED** [review]
I'd like to take another pass at our naming before merging but this is looking pretty good.

File: api/scpca_portal/models/sample.py [hunk 6]
```diff
@@ -56,25 +56,24 @@ def __str__(self):
     def get_from_dict(cls, data, project):
         """Prepares ready for saving sample object."""
 
-        has_multiplexed_data = bool(data.get("multiplexed_with"))
         sample = cls(
             age_at_diagnosis=data["age_at_diagnosis"],
-            demux_cell_count_estimate=(
-                data.get("demux_cell_count_estimate") if has_multiplexed_data else None
-            ),
+            demux_cell_count_estimate=(data.get("demux_cell_count_estimate", None)),
```
[the remark this answers is not in the document]
Comment (avrohomgottlieb) [reply api/scpca_portal/models/sample.py:61]: I believe so. The key is dropped because `demux_cell_count_estimate` is not included in the output metadata file, but we keep the property as an attribute on `Sample` objects. 
- When building `Sample` objects we use `updated_samples_metadata`.
- When combining metadata and writing the metadata files, we filter `updated_samples_metadata` and use  `samples_metadata_filtered_keys`.

File: api/scpca_portal/models/project.py (hunk 1 above)
Comment (avrohomgottlieb) [reply api/scpca_portal/models/project.py:493]: Good call, that's a better name. Will make the change

File: api/scpca_portal/models/project.py (hunk 4 above)
Comment (avrohomgottlieb) [reply api/scpca_portal/models/project.py:765]: My reasoning was that the normal case is *not* multiplexed, and the exception *is* multiplexed. Personal preference - I tend to prefer the rule first, and then the exception afterwards. Does the following help at all?
```python
is_not_multiplexed = "demux_samples" not in library_metadata

if is_not_multiplexed:
  ...
```

File: api/scpca_portal/models/project.py (hunk 5 above)
Comment (avrohomgottlieb) [reply api/scpca_portal/models/project.py:1216]: Sure. I actually originally used `input_samples_metadata`, but was afraid that the name was too generic and that it could be confused with the different usage of `input_samples_metadata` used elsewhere throughout in the file. I like `unfiltered_samples_metadata`.

File: api/scpca_portal/utils.py [hunk 7]
```diff
@@ -28,3 +28,15 @@ def join_workflow_versions(workflow_versions):
 def get_today_string(format: str = "%Y-%m-%d"):
     """Returns today's date formatted. Defaults to ISO 8601."""
     return datetime.today().strftime(format)
+
+
+def filter_dict_list_by_keys(list_of_dicts, included_keys):
```
[the remark this answers is not in the document]
Comment (avrohomgottlieb) [reply api/scpca_portal/utils.py:33]: Yes, we should.

[the remark this answers is not in the document]
Comment (avrohomgottlieb) [reply api/scpca_portal/utils.py]: Added.

**Review by davidsmejia — COMMENTED** [review]
Looking good, just a couple suggestions.

File: api/scpca_portal/utils.py [hunk 8]
```diff
@@ -28,3 +28,24 @@ def join_workflow_versions(workflow_versions):
 def get_today_string(format: str = "%Y-%m-%d"):
     """Returns today's date formatted. Defaults to ISO 8601."""
     return datetime.today().strftime(format)
+
+
+def filter_dict_list_by_keys(list_of_dicts, included_keys, ignore_value_error=False):
```
[the remark this answers is not in the document]
Comment (avrohomgottlieb) [reply api/scpca_portal/utils.py:33]: Ay ay captain!

File: api/scpca_portal/utils.py [hunk 9]
```diff
@@ -28,3 +28,24 @@ def join_workflow_versions(workflow_versions):
...
+    """
+    Returns a list of dictionaries with keys filtered according to provided include_keys.
+    If included-keys are not a subset of dictionary keys, returns a ValueError.
+    Throwing of ValueError can be disabled.
+    """
+    # Make sure included_keys is subset of dictionary keys
+    if not ignore_value_error and not set(included_keys).issubset(set(list_of_dicts[0].keys())):
```
[the remark this answers is not in the document]
Comment (avrohomgottlieb) [reply api/scpca_portal/utils.py:40]: This is a good point, I was thinking about it from the perspective of the current usage where all inputted sample metadata dictionaries have the same set of keys, but a utils function should have the expectation of being generic and having the ability to accept varied inputs.

Here's a better solution:
```python
if not ignore_value_error:
    if any(not set(included_keys).issubset(dictionary) for dictionary in list_of_dicts):
        raise ValueError("Included keys must be a subset of each dictionary's key set")
```

[the remark this answers is not in the document]
Comment (davidsmejia) [reply api/scpca_portal/utils.py]: Yeah this seems more reasonable.

**Review by davidsmejia — APPROVED** [review]
👍
